asset-transfer-basic/chaincode-java/src/main/java/org/hyperledger/fabric/samples/assetTransfer/AssetTransfer.java

```diff
@@ -0,0 +1,238 @@
+        }
+        
+        String assetState = stub.getStringState(assetID);
+        Asset asset = genson.deserialize(assetState, Asset.class);
+
+        Asset newAsset = new Asset(assetID, color, size, owner, appraisedValue);
+        String newAssetState = genson.serialize(newAsset);
+        stub.putStringState(assetID, newAssetState);
+
+        return newAsset;
+    }
+    
+    /**
+     * Deletes asset on the ledger.
+     *
+     * @param ctx the transaction context
+     * @param assetID the assetID
+     */
+    @Transaction()
+    public void DeleteAsset(final Context ctx, final String assetID) {
+        ChaincodeStub stub = ctx.getStub();
+
+        if (!AssetExists(ctx, assetID)) {
+            String errorMessage = String.format("Asset %s does not exist", assetID);
+            System.out.println(errorMessage);
+            throw new ChaincodeException(errorMessage, AssetTransferErrors.ASSET_NOT_FOUND.toString());
+        }
+
+        stub.delState(assetID);
+    }
+
+    /**
+     * Checks the existence of the asset on the ledger
+     *
+     * @param ctx the transaction context
+     * @param assetID the asset key
+     */
+    @Transaction()
+    public boolean AssetExists(final Context ctx, final String assetID) {
+        ChaincodeStub stub = ctx.getStub();
+        String assetState = stub.getStringState(assetID);
+
+        return (assetState != null && !assetState.isEmpty());
+
+    }
+
+    /**
+     * Changes the owner of a asset on the ledger.
+     *
+     * @param ctx the transaction context
+     * @param assetID the assetID
+     * @param newOwner the new owner
+     * @return the updated Asset
+     */
+    @Transaction()
+    public Asset TransferAsset(final Context ctx, final String assetID, final String newOwner) {
+        ChaincodeStub stub = ctx.getStub();
+
+        if (!AssetExists(ctx, assetID)) {
+            String errorMessage = String.format("Asset %s does not exist", assetID);
+            System.out.println(errorMessage);
+            throw new ChaincodeException(errorMessage, AssetTransferErrors.ASSET_NOT_FOUND.toString());
+        }
+        
+        String assetState = stub.getStringState(assetID);
+        Asset asset = genson.deserialize(assetState, Asset.class);
+
+        Asset newAsset = new Asset(asset.getAssetID(), asset.getColor(), asset.getSize(), newOwner, asset.getAppraisedValue());
+        String newAssetState = genson.serialize(newAsset);
+        stub.putStringState(assetID, newAssetState);
+
+        return newAsset;
+    }
+    
+    /**
+     * Retrieves all assets from the ledger.
+     *
+     * @param ctx the transaction context
+     * @return array of Assets found on the ledger
+     */
+    @Transaction()
+    public String GetAllAssets(final Context ctx) {
+        ChaincodeStub stub = ctx.getStub();
+
+        List<AssetQueryResult> queryResults = new ArrayList<AssetQueryResult>();
+
+        QueryResultsIterator<KeyValue> results = stub.getStateByRange("", "");
```

Comment:
Comment on this line, on the empty start/end keys, is helpful to new readers:
```
//To retrieve all assets from the ledger use getStateByRange with empty startKey & endKey. Giving empty startKey & endKey is interpreted as all the keys from beginning to end. 
//As another example, if you use startKey = 'asset0', endKey = 'asset9' , then getStateByRange will retrieve asset with keys between asset0 (inclusive) and asset9 (exclusive) in lexical order. 
```

Reply:
sure, I'll add that in